File: src/DataScience/Octave/factory_method.m
function factory_method()
  use_database(@create_postgres);
  use_database(@create_mysql);
end

function database = create_postgres()
  database.connect = @() disp('PostgreSQL connect');
  database.query = @() disp('PostgreSQL query');
end

function database = create_mysql()
  database.connect = @() disp('MySQL connect');
  database.query = @() disp('MySQL query');
end

function use_database(create_database)
  database = create_database();
  database.connect();
  database.query();
end

File: src/Niche/Nim/factory_method.nim
type
  DatabaseAction = proc()
  Database = object
    connect: DatabaseAction
    query: DatabaseAction
  FactoryMethod = proc(): Database

proc postgresConnect() = echo "PostgreSQL connect"
proc postgresQuery() = echo "PostgreSQL query"
proc mySqlConnect() = echo "MySQL connect"
proc mySqlQuery() = echo "MySQL query"

proc createPostgres(): Database =
  Database(connect: postgresConnect, query: postgresQuery)

proc createMySql(): Database =
  Database(connect: mySqlConnect, query: mySqlQuery)

proc useDatabase(createDatabase: FactoryMethod) =
  let database = createDatabase()
  database.connect()
  database.query()

useDatabase(createPostgres)
useDatabase(createMySql)
